math/vecops/inc/ROOT/RVec.hxx

```diff
@@ -680,14 +680,81 @@ RVec<typename RVec<T>::size_type> Argsort(const RVec<T> &v)
 
 /// Return elements of a vector at given indices
 template <typename T>
-RVec<T> ByIndices(const RVec<T> &v, const RVec<typename RVec<T>::size_type> &i)
+RVec<T> TakeIndices(const RVec<T> &v, const RVec<typename RVec<T>::size_type> &i)
 {
-   RVec<T> r(i.size());
-   for (unsigned int k = 0; k < i.size(); k++)
+
+   const auto vsize = v.size();
+   const auto isize = i.size();
+   if (isize > vsize) {
+      std::stringstream ss;
+      ss << "Try to take " << isize << " elements but vector has only size " << vsize << ".";
+      throw std::runtime_error(ss.str());
+   }
+   RVec<T> r(isize);
+   for (unsigned int k = 0; k < isize; k++)
       r[k] = v[i[k]];
    return r;
 }
 
+/// Return first elements of a vector
+template <typename T>
+RVec<T> TakeFirst(const RVec<T> &v, const typename RVec<T>::size_type n)
+{
+   const auto size = v.size();
+   if (n > size) {
+      std::stringstream ss;
+      ss << "Try to take " << n << " elements but vector has only size " << size << ".";
+      throw std::runtime_error(ss.str());
+   }
+   RVec<T> r(n);
+   for (typename RVec<T>::size_type k = 0; k < n; k++)
+      r[k] = v[k];
+   return r;
+}
+
+/// Return last elements of a vector
+template <typename T>
+RVec<T> TakeLast(const RVec<T> &v, const typename RVec<T>::size_type n)
+{
+   const auto size = v.size();
+   if (n > size) {
+      std::stringstream ss;
+      ss << "Try to take " << n << " elements but vector has only size " << size << ".";
+      throw std::runtime_error(ss.str());
+   }
+   RVec<T> r(n);
+   for (typename RVec<T>::size_type k = 0; k < n; k++)
+      r[k] = v[size - n + k];
+   return r;
+}
+
+/// Return copy of reversed vector
+template <typename T>
+RVec<T> Reverse(const RVec<T> &v)
+{
+   RVec<T> r(v);
+   std::reverse(r.begin(), r.end());
+   return r;
+}
+
+/// Return copy of vector with elements sorted in ascending order
+template <typename T>
+RVec<T> Sort(const RVec<T> &v)
+{
+   RVec<T> r(v);
+   std::sort(r.begin(), r.end());
+   return r;
+}
+
+/// Return copy of vector with elements sorted based on comparison operator
+template <typename T, typename Compare>
+RVec<T> Sort(const RVec<T> &v, Compare c)
+{
+   RVec<T> r(v);
+   std::sort(r.begin(), r.end(), std::forward<Compare>(c));
```

Comment:
@bluehood Do I understand correctly that the `std::forward` pushes as well the error-handling to the overload of `std::sort(begin, end, compare_thingy)`?

Reply:
Or can I do additional sanitizing? Are there any asserts I can do?

Reply:
for `std::forward` to be useful you should take a `Compare&&` forwarding reference.
regarding the checks, imo it's ok to rely on `std::sort` but then we should document that the requirements on `Compare` are the same as in `std::sort`

Reply:
done